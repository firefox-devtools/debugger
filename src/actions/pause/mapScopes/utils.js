// @flow

import type { Location } from "../../../types";

export function locColumn(loc: Location): number {
  if (typeof loc.column !== "number") {
    // This shouldn't really happen with locations from the AST, but
    // the datatype we are using allows null/undefined column.
    return 0;
  }

  return loc.column;
}
