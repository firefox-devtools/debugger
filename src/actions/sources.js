// @flow

/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * Redux actions for the sources state
 * @module actions/sources
 */

const defer = require("../utils/defer");
const { PROMISE } = require("../utils/redux/middleware/promise");
const assert = require("../utils/assert");
const { updateFrameLocations } = require("../utils/pause");
const {
  getOriginalURLs, getOriginalSourceText,
  generatedToOriginalId, isOriginalId,
  getOriginalLocation, getGeneratedLocation,
  isGeneratedId, applySourceMap, shouldSourceMap
} = require("../utils/source-map");

const { prettyPrint } = require("../utils/pretty-print");
const { getPrettySourceURL } = require("../utils/source");

const constants = require("../constants");
const { removeDocument } = require("../utils/source-documents");

const {
  getSource, getSourceByURL, getSourceText,
  getPendingSelectedLocation, getFrames
} = require("../selectors");

import type { Source } from "../types";
import type { ThunkArgs } from "./types";

/**
 * Handler for the debugger client's unsolicited newSource notification.
 * @memberof actions/sources
 * @static
 */
function newSource(source: Source) {
  return ({ dispatch, getState }: ThunkArgs) => {
    if (shouldSourceMap()) {
      dispatch(loadSourceMap(source));
    }

    dispatch({
      type: constants.ADD_SOURCE,
      source
    });

    // If a request has been made to show this source, go ahead and
    // select it.
    const pendingLocation = getPendingSelectedLocation(getState());
    if (pendingLocation && pendingLocation.url === source.url) {
      dispatch(selectSource(source.id, { line: pendingLocation.line }));
    }
  };
}

function newSources(sources: Source[]) {
  return ({ dispatch, getState }: ThunkArgs) => {
    sources.filter(source => !getSource(getState(), source.id))
      .forEach(source => dispatch(newSource(source)));
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function loadSourceMap(generatedSource) {
  return async function({ dispatch, getState }: ThunkArgs) {
    const urls = await getOriginalURLs(generatedSource);
    if (!urls) {
      // If this source doesn't have a sourcemap, do nothing.
      return;
    }

    const originalSources = urls.map(originalUrl => {
      return {
        url: originalUrl,
        id: generatedToOriginalId(generatedSource.id, originalUrl),
        isPrettyPrinted: false
      };
    });

    originalSources.forEach(s => dispatch(newSource(s)));
  };
}

type SelectSourceOptions = {
  tabIndex?: number,
  line?: number
};

/**
 * Deterministically select a source that has a given URL. This will
 * work regardless of the connection status or if the source exists
 * yet. This exists mostly for external things to interact with the
 * debugger.
 *
 * @memberof actions/sources
 * @static
 */
function selectSourceURL(url: string, options: SelectSourceOptions = {}) {
  return ({ dispatch, getState }: ThunkArgs) => {
    const source = getSourceByURL(getState(), url);
    if (source) {
      dispatch(selectSource(source.get("id"), options));
    } else {
      dispatch({
        type: constants.SELECT_SOURCE_URL,
        url: url,
        tabIndex: options.tabIndex,
        line: options.line
      });
    }
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function selectSource(id: string, options: SelectSourceOptions = {}) {
  return ({ dispatch, getState, client }: ThunkArgs) => {
    if (!client) {
      // No connection, do nothing. This happens when the debugger is
      // shut down too fast and it tries to display a default source.
      return;
    }

    const source = getSource(getState(), id).toJS();

    // Make sure to start a request to load the source text.
    dispatch(loadSourceText(source));

    dispatch({ type: constants.TOGGLE_FILE_SEARCH, searchOn: false });

    dispatch({
      type: constants.SELECT_SOURCE,
      source: source,
      tabIndex: options.tabIndex,
      line: options.line
    });
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function jumpToMappedLocation(sourceLocation: any) {
  return async function({ dispatch, getState, client }: ThunkArgs) {
    if (!client) {
      return;
    }

    const source = getSource(getState(), sourceLocation.sourceId);
    const pairedLocation = isOriginalId(sourceLocation.sourceId)
      ? await getGeneratedLocation(sourceLocation, source.toJS())
      : await getOriginalLocation(sourceLocation, source.toJS());

    return dispatch(selectSource(
      pairedLocation.sourceId,
      { line: pairedLocation.line }
    ));
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function closeTab(url: string) {
  removeDocument(url);
  return {
    type: constants.CLOSE_TAB,
    url
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function closeTabs(urls: string[]) {
  return ({ dispatch, getState, client }: ThunkArgs) => {
    urls.forEach(url => {
      const source = getSourceByURL(getState(), url);
      if (source) {
        removeDocument(source.get("id"));
      }
    });

    dispatch({
      type: constants.CLOSE_TABS,
      urls
    });
  };
}

/**
 * Toggle the pretty printing of a source's text. All subsequent calls to
 * |getText| will return the pretty-toggled text. Nothing will happen for
 * non-javascript files.
 *
 * @memberof actions/sources
 * @static
 * @param string id The source form from the RDP.
 * @returns Promise
 *          A promise that resolves to [aSource, prettyText] or rejects to
 *          [aSource, error].
 */
function togglePrettyPrint(sourceId: string) {
  return ({ dispatch, getState, client }: ThunkArgs) => {
    const source = getSource(getState(), sourceId).toJS();
    const sourceText = getSourceText(getState(), sourceId).toJS();

    if (sourceText.loading) {
      return {};
    }

    assert(isGeneratedId(sourceId),
           "Pretty-printing only allowed on generated sources");

    const url = getPrettySourceURL(source.url);
    const id = generatedToOriginalId(source.id, url);
    const originalSource = { url, id, isPrettyPrinted: false };
    dispatch({
      type: constants.ADD_SOURCE,
      source: originalSource
    });

    return dispatch({
      type: constants.TOGGLE_PRETTY_PRINT,
      source: originalSource,
      [PROMISE]: (async function() {
        const { code, mappings } = await prettyPrint({
          source, sourceText, url
        });
        await applySourceMap(source.id, url, code, mappings);

        const frames = await updateFrameLocations(getFrames(getState()));
        dispatch(selectSource(originalSource.id));

        return {
          text: code,
          contentType: "text/javascript",
          frames
        };
      })()
    });
  };
}

/**
 * @memberof actions/sources
 * @static
 */
function loadSourceText(source: Source) {
  return ({ dispatch, getState, client }: ThunkArgs) => {
    // Fetch the source text only once.
    let textInfo = getSourceText(getState(), source.id);
    if (textInfo) {
      // It's already loaded or is loading
      return Promise.resolve(textInfo);
    }

    return dispatch({
      type: constants.LOAD_SOURCE_TEXT,
      source: source,
      [PROMISE]: (async function() {
        if (isOriginalId(source.id)) {
          return await getOriginalSourceText(source);
        }

        const response = await client.sourceContents(source.id);
        return {
          text: response.source,
          contentType: response.contentType || "text/javascript"
        };

        // Automatically pretty print if enabled and the test is
        // detected to be "minified"
        // if (Prefs.autoPrettyPrint &&
        //     !source.isPrettyPrinted &&
        //     SourceUtils.isMinified(source.id, response.source)) {
        //   dispatch(togglePrettyPrint(source));
        // }
      })()
    });
  };
}

// delay is in ms
const FETCH_SOURCE_RESPONSE_DELAY = 200;

/**
 * Starts fetching all the sources, silently.
 *
 * @memberof actions/sources
 * @static
 * @param array actors
 *        The urls for the sources to fetch. If fetching a source's text
 *        takes too long, it will be discarded.
 * @returns {Promise}
 *         A promise that is resolved after source texts have been fetched.
 */
function getTextForSources(actors: any[]) {
  return ({ dispatch, getState }: ThunkArgs) => {
    let deferred = defer();
    let pending = new Set(actors);
    type FetchedSourceType = [any, string, string];
    let fetched: FetchedSourceType[] = [];

    // Can't use promise.all, because if one fetch operation is rejected, then
    // everything is considered rejected, thus no other subsequent source will
    // be getting fetched. We don't want that. Something like Q's allSettled
    // would work like a charm here.

    // Try to fetch as many sources as possible.
    for (let actor of actors) {
      let source = getSource(getState(), actor);
      dispatch(loadSourceText(source)).then(({ text, contentType }) => {
        onFetch([source, text, contentType]);
      }, err => {
        onError(source, err);
      });
    }

    setTimeout(onTimeout, FETCH_SOURCE_RESPONSE_DELAY);

    /* Called if fetching a source takes too long. */
    function onTimeout() {
      pending = new Set();
      maybeFinish();
    }

    /* Called if fetching a source finishes successfully. */
    function onFetch([aSource, aText, aContentType]: FetchedSourceType) {
      // If fetching the source has previously timed out, discard it this time.
      if (!pending.has(aSource.actor)) {
        return;
      }
      pending.delete(aSource.actor);
      fetched.push([aSource.actor, aText, aContentType]);
      maybeFinish();
    }

    /* Called if fetching a source failed because of an error. */
    function onError([aSource, aError]) {
      pending.delete(aSource.actor);
      maybeFinish();
    }

    /* Called every time something interesting
     *  happens while fetching sources.
     */
    function maybeFinish() {
      if (pending.size == 0) {
        // Sort the fetched sources alphabetically by their url.
        if (deferred) {
          deferred.resolve(
            fetched.sort(([aFirst], [aSecond]) => aFirst > aSecond ? -1 : 1)
          );
        }
      }
    }

    return deferred.promise;
  };
}

module.exports = {
  newSource,
  newSources,
  selectSource,
  selectSourceURL,
  jumpToMappedLocation,
  closeTab,
  closeTabs,
  togglePrettyPrint,
  loadSourceText,
  getTextForSources
};
