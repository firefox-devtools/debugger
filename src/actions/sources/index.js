export { loadSourceText } from "./loadSourceText";
export { togglePrettyPrint } from "./prettyPrint";
