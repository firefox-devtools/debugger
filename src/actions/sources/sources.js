/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import type { Action } from "../types";

export const setSelectedLocation: Action = (source, location) => ({
  type: "SET_SELECTED_LOCATION",
  source,
  location
});

export const clearSelectedSource: Action = () => ({
  type: "CLEAR_SELECTED_SOURCE"
});
