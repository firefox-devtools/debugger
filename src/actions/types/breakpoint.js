// @flow

import type { Breakpoint, Location } from "../../types";

import type { PromiseAction } from "../utils/middleware/promise";

type AddBreakpointResult = {
  previousLocation: Location,
  breakpoint: Breakpoint
};

export type BreakpointAction =
  | PromiseAction<
      {|
        +type: "ADD_BREAKPOINT",
        +breakpoint: Breakpoint,
        +condition?: string
      |},
      AddBreakpointResult
    >
  | PromiseAction<{|
      +type: "REMOVE_BREAKPOINT",
      +breakpoint: Breakpoint,
      +disabled: boolean
    |}>
  // for simulating a successful server request
  | {|
      +type: "REMOVE_BREAKPOINT",
      +breakpoint: Breakpoint,
      +status: "done"
    |}
  | {|
      +type: "SET_BREAKPOINT_CONDITION",
      +breakpoint: Breakpoint
    |}
  | PromiseAction<{|
      +type: "TOGGLE_BREAKPOINTS",
      +shouldDisableBreakpoints: boolean
    |}>
  | {|
      +type: "SYNC_BREAKPOINT",
      +breakpoint: ?Breakpoint,
      +previousLocation: Location
    |}
  | PromiseAction<
      {|
        +type: "ENABLE_BREAKPOINT",
        +breakpoint: Breakpoint
      |},
      AddBreakpointResult
    >
  | {|
      +type: "DISABLE_BREAKPOINT",
      +breakpoint: Breakpoint
    |}
  | {|
      +type: "DISABLE_ALL_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |}
  | {|
      +type: "ENABLE_ALL_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |}
  | {|
      +type: "REMAP_BREAKPOINTS",
      +breakpoints: Breakpoint[]
    |};
