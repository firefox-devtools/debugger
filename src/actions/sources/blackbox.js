/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

/**
 * Redux actions for the sources state
 * @module actions/sources
 */

import { isOriginalId } from "devtools-source-map";
import { recordEvent } from "../../utils/telemetry";
import { features } from "../../utils/prefs";

import { PROMISE } from "../utils/middleware/promise";
import type { Source } from "../../types";
import type { ThunkArgs } from "../types";

export function toggleBlackBox(source: Source) {
  return async ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
    const { isBlackBoxed, id } = source;

    if (!isBlackBoxed) {
      recordEvent("blackbox");
    }

    let promise;
    if (features.originalBlackbox && isOriginalId(id)) {
      promise = Promise.resolve({isBlackBoxed: !isBlackBoxed})
    } else {
      promise = client.blackBox(id, isBlackBoxed)
    }

    return dispatch({
      type: "BLACKBOX",
      source,
      [PROMISE]: promise
    });
  };
}
