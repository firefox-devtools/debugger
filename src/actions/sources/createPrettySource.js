import { getFrames, getSource } from "../../selectors";
import { prettyPrint } from "../../workers/pretty-print";
import { updateFrameLocations } from "../../utils/pause";
import { getPrettySourceURL } from "../../utils/source";

export function createPrettySource(sourceId) {
  return async ({ dispatch, getState, sourceMaps }) => {
    const source = getSource(getState(), sourceId).toJS();
    const url = getPrettySourceURL(source.url);
    const id = await sourceMaps.generatedToOriginalId(sourceId, url);

    const { code, mappings } = await prettyPrint({
      source,
      url
    });

    await sourceMaps.applySourceMap(source.id, url, code, mappings);

    let frames = getFrames(getState());
    if (frames) {
      frames = await updateFrameLocations(frames, sourceMaps);
    }

    const prettySource = {
      url,
      id,
      isPrettyPrinted: true,
      text: code,
      contentType: "text/javascript",
      frames,
      loadedState: "loaded"
    };

    dispatch({
      type: "ADD_SOURCE",
      source: prettySource
    });

    return prettySource;
  };
}
