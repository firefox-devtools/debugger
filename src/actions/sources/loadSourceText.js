/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { PROMISE } from "../utils/middleware/promise";
import { getGeneratedSource, getSource } from "../../selectors";
import * as parser from "../../workers/parser";
import { isLoaded, isOriginal } from "../../utils/source";
import { Telemetry } from "devtools-modules";

import defer from "../../utils/defer";
import type { ThunkArgs } from "../types";

import type { Source } from "../../types";

const requests = new Map();

// Measures the time it takes for a source to load
const loadSourceHistogram = "DEVTOOLS_DEBUGGER_LOAD_SOURCE_MS";
const telemetry = new Telemetry();

async function loadSource(source: Source, { sourceMaps, client }) {
  const { id } = source;
  if (isOriginal(source)) {
    return sourceMaps.getOriginalSourceText(source);
  }

  const response = await client.sourceContents(id);
  telemetry.finish(loadSourceHistogram, source);

  return {
    id,
    text: response.source,
    contentType: response.contentType || "text/javascript"
  };
}

/**
 * @memberof actions/sources
 * @static
 */
export function loadSourceText(source: ?Source) {
  return async ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
    if (!source) {
      return;
    }

    const id = source.id;
    // Fetch the source text only once.
    if (requests.has(id)) {
      return requests.get(id);
    }

    if (isLoaded(source)) {
      return Promise.resolve();
    }

    const deferred = defer();
    requests.set(id, deferred.promise);

    telemetry.start(loadSourceHistogram, source);
    try {
      await dispatch({
        type: "LOAD_SOURCE_TEXT",
        sourceId: source.id,
        [PROMISE]: loadSource(source, { sourceMaps, client })
      });
    } catch (e) {
      deferred.resolve();
      requests.delete(id);
      return;
    }

    const newSource = getSource(getState(), source.id);
    if (!newSource) {
      return;
    }

    if (isOriginal(newSource) && !newSource.isWasm) {
      const generatedSource = getGeneratedSource(getState(), source);
      await dispatch(loadSourceText(generatedSource));
    }

    if (!newSource.isWasm) {
      await parser.setSource(newSource);
    }

    // signal that the action is finished
    deferred.resolve();
    requests.delete(id);

    return source;
  };
}
