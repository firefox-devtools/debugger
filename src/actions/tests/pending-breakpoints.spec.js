// TODO: we would like to mock this in the local tests
import {
  generateBreakpoint,
  mockPendingBreakpoint
} from "./helpers/breakpoints.js";

import {
  simulateCorrectThreadClient,
  simpleMockThreadClient
} from "./helpers/threadClient.js";

import { prefs } from "../../utils/prefs";

jest.mock("../../utils/prefs", () => ({
  prefs: {
    expressions: [],
    pendingBreakpoints: {}
  },
  clear: jest.fn()
}));

jest.mock("../sources/loadSourceText", () => ({
  loadSourceText: () => ({
    type: "LOAD_SOURCE_TEXT",
    source: {
      id: "id",
      text: "function foo() {}",
      contentType: "text/javascript"
    },
    value: {
      id: "id",
      text: "function () {}",
      contentType: "text/javascript"
    }
  })
}));
import "../sources/loadSourceText";

import {
  createStore,
  selectors,
  actions,
  makeSource
} from "../../utils/test-head";

import { makePendingLocationId, makeLocationId } from "../../utils/breakpoint";

describe("when adding breakpoints", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint();

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("a corresponding pending breakpoint should be added", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);

    await dispatch(actions.newSource(makeSource("foo.js")));

    const bp = generateBreakpoint("foo.js");
    const id = makePendingLocationId(bp.location);

    await dispatch(actions.addBreakpoint(bp.location));
    const pendingBps = selectors.getPendingBreakpoints(getState());
    expect(pendingBps.size).toBe(2);
    expect(pendingBps.get(id)).toMatchSnapshot();
  });

  describe("adding and deleting breakpoints", () => {
    let breakpoint1;
    let breakpoint2;
    let breakpointLocationId1;
    let breakpointLocationId2;

    beforeEach(() => {
      breakpoint1 = generateBreakpoint("foo");
      breakpoint2 = generateBreakpoint("foo2");
      breakpointLocationId1 = makePendingLocationId(breakpoint1.location);
      breakpointLocationId2 = makePendingLocationId(breakpoint2.location);
    });

    it("add a corresponding pendingBreakpoint for each addition", async () => {
      const { dispatch, getState } = createStore(simpleMockThreadClient);

      await dispatch(actions.newSource(makeSource("foo")));
      await dispatch(actions.newSource(makeSource("foo2")));

      await dispatch(actions.addBreakpoint(breakpoint1.location));
      await dispatch(actions.addBreakpoint(breakpoint2.location));

      const pendingBps = selectors.getPendingBreakpoints(getState());
      expect(pendingBps.get(breakpointLocationId1)).toMatchSnapshot();
      expect(pendingBps.get(breakpointLocationId2)).toMatchSnapshot();
    });

    it("hidden breakponts do not create pending bps", async () => {
      const { dispatch, getState } = createStore(simpleMockThreadClient);

      await dispatch(actions.newSource(makeSource("foo")));
      await dispatch(
        actions.addBreakpoint(breakpoint1.location, { hidden: true })
      );
      const pendingBps = selectors.getPendingBreakpoints(getState());

      expect(pendingBps.get(breakpointLocationId1)).toBeUndefined();
    });

    it("remove a corresponding pending breakpoint when deleting", async () => {
      const { dispatch, getState } = createStore(simpleMockThreadClient);
      await dispatch(actions.newSource(makeSource("foo")));
      await dispatch(actions.newSource(makeSource("foo2")));

      await dispatch(actions.addBreakpoint(breakpoint1.location));
      await dispatch(actions.addBreakpoint(breakpoint2.location));
      await dispatch(actions.removeBreakpoint(breakpoint1.location));

      const pendingBps = selectors.getPendingBreakpoints(getState());
      expect(pendingBps.has(breakpointLocationId1)).toBe(false);
      expect(pendingBps.has(breakpointLocationId2)).toBe(true);
    });
  });
});

describe("when changing an existing breakpoint", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint();

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("updates corresponding pendingBreakpoint", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);
    const bp = generateBreakpoint("foo");
    const id = makePendingLocationId(bp.location);
    await dispatch(actions.newSource(makeSource("foo")));

    await dispatch(actions.addBreakpoint(bp.location));
    await dispatch(
      actions.setBreakpointCondition(bp.location, { condition: "2" })
    );
    const bps = selectors.getPendingBreakpoints(getState());
    const breakpoint = bps.get(id);
    expect(breakpoint.condition).toBe("2");
  });

  it("if disabled, updates corresponding pendingBreakpoint", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);
    const bp = generateBreakpoint("foo");
    const id = makePendingLocationId(bp.location);

    await dispatch(actions.newSource(makeSource("foo")));
    await dispatch(actions.addBreakpoint(bp.location));
    await dispatch(actions.disableBreakpoint(bp.location));
    const bps = selectors.getPendingBreakpoints(getState());
    const breakpoint = bps.get(id);
    expect(breakpoint.disabled).toBe(true);
  });

  it("does not delete the pre-existing pendingBreakpoint", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);
    const bp = generateBreakpoint("foo.js");
    const source = makeSource("foo.js");
    await dispatch(actions.newSource(source));

    const id = makePendingLocationId(bp.location);

    await dispatch(actions.addBreakpoint(bp.location));
    await dispatch(
      actions.setBreakpointCondition(bp.location, { condition: "2" })
    );
    const bps = selectors.getPendingBreakpoints(getState());
    const breakpoint = bps.get(id);
    expect(breakpoint.condition).toBe("2");
  });
});

describe("initializing when pending breakpoints exist in prefs", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint();

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("syncs pending breakpoints", async () => {
    const { getState } = createStore(simpleMockThreadClient);
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    const bps = selectors.getPendingBreakpoints(getState());
    const bp = bps.get(id);
    expect(bp).toMatchSnapshot();
  });

  it("re-adding breakpoints update existing pending breakpoints", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);
    const bar = generateBreakpoint("bar.js");
    await dispatch(actions.newSource(makeSource("bar.js")));
    await dispatch(actions.addBreakpoint(bar.location));

    const bps = selectors.getPendingBreakpoints(getState());
    expect(bps.size).toBe(1);
  });

  it("adding bps doesn't remove existing pending breakpoints", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);
    const bp = generateBreakpoint("foo.js");

    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.addBreakpoint(bp.location));

    const bps = selectors.getPendingBreakpoints(getState());
    expect(bps.size).toBe(2);
  });
});

describe("initializing with disabled pending breakpoints in prefs", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint({ disabled: true });

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("syncs breakpoints with pending breakpoints", async () => {
    const expectedLocation = Object.assign(
      {},
      mockedPendingBreakpoint.location,
      { sourceId: "bar.js" }
    );

    const expectedId = makeLocationId(expectedLocation);
    const { getState, dispatch } = createStore(simpleMockThreadClient);
    const source = makeSource("bar.js");
    await dispatch(actions.newSource(source));
    const bps = selectors.getBreakpoints(getState());
    const bp = bps.get(expectedId);

    expect(bp.location).toEqual(expectedLocation);
    expect(bp.disabled).toEqual(mockedPendingBreakpoint.disabled);
  });
});

describe("adding sources", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint();

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("corresponding breakpoints are added for a single source", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);

    let bps = selectors.getBreakpoints(getState());
    expect(bps.size).toBe(0);

    const source = makeSource("bar.js");
    await dispatch(actions.newSource(source));
    bps = selectors.getBreakpoints(getState());
    expect(bps.size).toBe(1);
  });

  it("add corresponding breakpoints for multiple sources", async () => {
    const { dispatch, getState } = createStore(simpleMockThreadClient);

    let bps = selectors.getBreakpoints(getState());
    expect(bps.size).toBe(0);

    const source1 = makeSource("bar.js");
    const source2 = makeSource("foo.js");
    await dispatch(actions.newSources([source1, source2]));
    bps = selectors.getBreakpoints(getState());
    expect(bps.size).toBe(1);
  });
});

describe("invalid breakpoint location", () => {
  const mockedPendingBreakpoint = mockPendingBreakpoint();

  beforeEach(() => {
    const id = makePendingLocationId(mockedPendingBreakpoint.location);
    prefs.pendingBreakpoints = { [id]: mockedPendingBreakpoint };
  });

  it("a corrected corresponding pending breakpoint is added", async () => {
    // setup
    const bp = generateBreakpoint("foo.js");
    const {
      correctedThreadClient,
      correctedLocation
    } = simulateCorrectThreadClient(2, bp.location);
    const { dispatch, getState } = createStore(correctedThreadClient);
    const correctedPendingId = makePendingLocationId(correctedLocation);

    // test
    await dispatch(actions.newSource(makeSource("foo.js")));
    await dispatch(actions.addBreakpoint(bp.location));
    const pendingBps = selectors.getPendingBreakpoints(getState());
    const pendingBp = pendingBps.get(correctedPendingId);
    expect(pendingBp).toMatchSnapshot();
  });
});
