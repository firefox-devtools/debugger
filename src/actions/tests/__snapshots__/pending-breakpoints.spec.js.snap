// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializing when pending breakpoints exist in prefs syncs pending breakpoints 1`] = `
Object {
  "astLocation": Object {
    "name": undefined,
    "offset": Object {
      "line": 5,
    },
  },
  "condition": null,
  "disabled": false,
  "generatedLocation": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/bar.js",
  },
  "hidden": false,
  "location": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/bar.js",
  },
}
`;

exports[`invalid breakpoint location a corrected corresponding pending breakpoint is added 1`] = `
Object {
  "astLocation": Object {
    "name": undefined,
    "offset": Object {
      "line": 5,
      "sourceId": "foo.js",
      "sourceUrl": "http://localhost:8000/examples/foo.js",
    },
  },
  "condition": null,
  "disabled": false,
  "generatedLocation": Object {
    "column": undefined,
    "line": 7,
    "sourceUrl": "http://localhost:8000/examples/foo.js",
  },
  "location": Object {
    "column": undefined,
    "line": 7,
    "sourceUrl": "http://localhost:8000/examples/foo.js",
  },
}
`;

exports[`when adding breakpoints a corresponding pending breakpoint should be added 1`] = `
Object {
  "astLocation": Object {
    "name": undefined,
    "offset": Object {
      "line": 5,
      "sourceId": "foo.js",
      "sourceUrl": "http://localhost:8000/examples/foo.js",
    },
  },
  "condition": null,
  "disabled": false,
  "generatedLocation": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo.js",
  },
  "location": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo.js",
  },
}
`;

exports[`when adding breakpoints adding and deleting breakpoints add a corresponding pendingBreakpoint for each addition 1`] = `
Object {
  "astLocation": Object {
    "name": undefined,
    "offset": Object {
      "line": 5,
      "sourceId": "foo",
      "sourceUrl": "http://localhost:8000/examples/foo",
    },
  },
  "condition": null,
  "disabled": false,
  "generatedLocation": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo",
  },
  "location": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo",
  },
}
`;

exports[`when adding breakpoints adding and deleting breakpoints add a corresponding pendingBreakpoint for each addition 2`] = `
Object {
  "astLocation": Object {
    "name": undefined,
    "offset": Object {
      "line": 5,
      "sourceId": "foo2",
      "sourceUrl": "http://localhost:8000/examples/foo2",
    },
  },
  "condition": null,
  "disabled": false,
  "generatedLocation": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo2",
  },
  "location": Object {
    "column": undefined,
    "line": 5,
    "sourceUrl": "http://localhost:8000/examples/foo2",
  },
}
`;
